shared/templates/firefox_policy-setting/tests/correct_value.pass.sh

```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+firefox_cfg="policies.json"
+firefox_dirs="/usr/lib/firefox/distribution /usr/lib64/firefox/distribution /usr/local/lib/firefox/distribution /usr/local/lib64/firefox/distribution"
+
+{{{ find_python() }}}
+
+# Iterate over the possible Firefox install directories
+for firefox_dir in ${firefox_dirs}; do
+    # write our bad config in every location to ensure remediation fixes it.
+    mkdir -p ${firefox_dir}
+    echo "{ \"policies\" :{ } }" > ${firefox_dir}/${firefox_cfg}
+    echo """
+import json
+_file=open('${firefox_dir}/${firefox_cfg}', 'r')
```

Comment:
Since I was using a Python snippet anyway, it made sense at thee time to re-use that embedded Python script so that I knew _a priori_ that it would be valid. 

Reimplementing with "just `bash` and `jinja`" means that all of the python here gets pushed up to the template process function and the entirety of the file write gets done as a block `cat << EOF  > ${firefox_dir}/${firefox_cfg} ... stuff EOF` and the resulting JSON is still done in Python.

Since it's just tests, if we had buy-in from the people maintaining all the build/test infrastructure to ensure `jq` is available and in PATH (and amended the docs so that other poor souls trying to run tests on their own), these could leverage `jq`.

Reply:
Yeah, I went and triple checked and `jq` is an EPEL package, which makes it unobtanium in my use case.

Reply:
So I spent some time pushing the actual test output generation up into `template.py` and that work actually makes the test harder to debug/review, because of the way the templates interact with tests, especially in a container.

I had to stick a `print(filled_templates)` statement into `ssg/templates.py` under `get_all_tests()` so that I could actually see and verify my text output and realize the refactoring mistake I had made. It isn't pretty *at all*.

Reply:
A here-document doesn't help when I still have to *write the here-document*, and that's non-trivial when I'm working *in the template*. I did write Python code to generate that here-document, but there's nowhere I can put that other than the Python template code. 

```python
def build_test_json(policy, _test, missing=False, wrong=False):
    """Generate a minimal JSON configuration for this particular policy object, based 
       on a correct (set to value), missing, or wrong value. 
       Returns a dict() containing this test entry.
       
       Parameters:
           policy: current policy item to generate for
           _test: dict() that has at least the value {"policies": {} }
    """
    _current = _test.get("policies")
    _key =  policy["parameter"]
    _value = make_json_boolean(policy["value"])

    for p in policy["subpath"]:
        if p not in _current:
           _current[p] = dict()
        _current = _current.get(p)
    
    if wrong:
        _current[_key] = "VERYVERYBADVALUE"
    elif not missing:
        _current[_key] = _value
    return _test
```

What I don't want is to spread the tests of this template across 10-20 individual rules; duplicating effort and separating the test from the implementing code. 
Otherwise I wouldn't have bothered with the template (which was the approach of the original writer of the lock parameter rules) in the first place.

This is the same approach (to keep the tests of the template with the template) as I took for a template for the [firefox_lockparameter](https://github.com/ComplianceAsCode/content/tree/master/shared/templates/firefox_lockparameter/tests) template and is also done in other places like `audit_rules_unsuccessful_file_modification`.

Reply:
I think that this is a step in a wrong direction. Previously, you had the Python code in the test scenario template, and  I think that you can actually generate the json file directly and easily at the same time. It also feels better to have different code for test scenario setup and the remediation.

Our test suite lacks a possibility to view a test scenario code after its jinja-expansion, which makes this quite annoying. I will look into it though.

Reply:
This code *does* generate the json and makes it a string that is then used for a here-document. Unless you want to make `jq` a hard dependency to run these tests, there's not a solution that is going to work other than either what I did here (generating the json during template instantiation/preprocessing), generating bespoke JSON for every rule that uses the template and placing those with the rule documents, or going down the path of madness by trying to assemble valid JSON with just standard POSIX tools. 

Personally, I think that either pushing the "here-document generation" step into the preprocess or leveraging the mechanism the remediation used is the less complex manner that doesn't give up the benefits of having the template in the first place.

Using `jq` is not something I can decide, given it's non-existence in RHEL standard repos. 

Generating bespoke JSON is a route I am going to resist as it introduces more complexity to implementing rules for this STIG in the future (and more places to screw up). 

The path of madness is right out, as I have already spent far too much effort trying to do that (which led to using Python as a stand-in for `jq` in the first place).

>Our test suite lacks a possibility to view a test scenario code after its jinja-expansion, which makes this quite annoying. I will look into it though.

When I was checking this code I had stuck a print statement for the expanded templates in that function so that I could tell what was going on (because I was running in a container, which further obfuscated things). Being able to dump those expanded scripts to the log directory would be extremely useful.